src/Nethermind/Nethermind.Trie/NibbleExtensions.cs

```diff
@@ -18,26 +21,68 @@ public static Nibble[] FromBytes(params byte[] bytes)
         public static Nibble[] FromBytes(ReadOnlySpan<byte> bytes)
         {
             Nibble[] nibbles = new Nibble[2 * bytes.Length];
-            for (int i = 0; i < bytes.Length; i++)
-            {
-                nibbles[i * 2] = new Nibble((byte)((bytes[i] & 240) >> 4));
-                nibbles[i * 2 + 1] = new Nibble((byte)(bytes[i] & 15));
-            }
-
+            BytesToNibbleBytes(bytes, MemoryMarshal.AsBytes(nibbles.AsSpan()));
             return nibbles;
         }
 
-        public static void BytesToNibbleBytes(ReadOnlySpan<byte> bytes, Span<byte> nibbles)
+        public unsafe static void BytesToNibbleBytes(ReadOnlySpan<byte> bytes, Span<byte> nibbles)
         {
+            // Ensure the length of the nibbles span is exactly twice the length of the bytes span.
             if (nibbles.Length != 2 * bytes.Length)
             {
                 ThrowArgumentException();
             }
 
-            for (int i = 0; i < bytes.Length; i++)
+            // Calculate the length to process using SIMD operations.
+            var length = bytes.Length / sizeof(Vector128<byte>) * sizeof(Vector128<byte>);
+            // Check if SIMD hardware acceleration is available and if there is data to process.
+            // This will be branch eliminated the asm if not supported.
+            if (Vector128.IsHardwareAccelerated && length > 0)
+            {
+                // Cast the byte span to a span of Vector128<byte> for SIMD processing.
+                var input = MemoryMarshal.Cast<byte, Vector128<byte>>(bytes.Slice(0, length));
+                // Cast the nibble span to a reference to first element of Vector128<ushort> as input doubles.
+                ref var output = ref Unsafe.As<byte, Vector128<ushort>>(ref MemoryMarshal.GetReference(nibbles));
+
+                for (int i = 0; i < input.Length; i++)
+                {
+                    // Get the bytes where each byte contains 2 nibbles that we want to move into their own byte.
+                    Vector128<byte> value = input[i];
+
+                    // Mask off lower nibble 0x0f and split each byte into two bytes and store them in two separate vectors.
+                    (Vector128<ushort> lower0, Vector128<ushort> upper0) = Vector128.Widen(Vector128.BitwiseAnd(value, Vector128.Create((byte)0x0f)));
+                    // Arrange the 0x0f nibbles; we use the 1st element to represent 0s, and then order as 0,2,4,6,8,10,12,14th elements.
+                    // This leaves byte holes for the other set of nibbles to fill.
+                    lower0 = Vector128.Shuffle(lower0.AsByte(), Vector128.Create((byte)1, 0, 1, 2, 1, 4, 1, 6, 1, 8, 1, 10, 1, 12, 1, 14)).AsUInt16();
+                    upper0 = Vector128.Shuffle(upper0.AsByte(), Vector128.Create((byte)1, 0, 1, 2, 1, 4, 1, 6, 1, 8, 1, 10, 1, 12, 1, 14)).AsUInt16();
```

Comment:
Those Vector.Create should be saved as statics?

Reply:
Yep; Jit moves them out of loop and loads from readonly data section of dll.

I tried to optimize it manually but ended up making worse asm than the Jit does like this